README.md
==============================================================================
ReadMe file for the swingset-parent folder.
==============================================================================

When SwingSet was converted to a Maven project, SwingSet and it's Demo
were split into two separate modules, swingset and swingset-demo.

Each module has a fully independent pom.xml, but there is also a
swingset-parent artifact and pom.xml that can be used to manage/compile both
modules.

Unfortunately, for the most seamless integration with Eclipse, Eclipse
wants to treat swingset-parent, swingset, and swingset-demo as three
separate projects.

To view as a single project:
 1. Window->Show View->Project Explorer
 2. View Menu (down arrow/triangle)->Project Presentation->Hierarchical
 
 Detailed ReadMe files for swingset and swingset-demo can be found in
 their respective folders.

ChangeLog.txt
==============================================================================
ChangeLog file for the SwingSet Open Toolkit for Java Swing.
==============================================================================

==============================================================================
SwingSet 3.0.0 - Released TBD
==============================================================================

Changes:
	1. Changed package from com.nqadmin.swingSet to com.nqadmin.swingset
	
==============================================================================
SwingSet 2.3.1 - Released 2019-11-11
==============================================================================

Changes:
	1. SSDataNavigator.setSSRowSet() only enabling add/delete if this.modification==true
	
==============================================================================
SwingSet 2.3.0 - Released 2019-11-09
==============================================================================

Changes:
	1. Removed BeanInfo and SSFormattedCombobBox classes (moved to archive)
	2. Addressed additional Eclipse compiler warnings
	3. Addressed additional Eclipse JavaDoc warnings
	4. Made many method parameters final
	5. Allowing Eclipse to format code when saving a class
	6. Removed any unqualified datamember access (e.g., use "this.x")
	7. Made inner classes protected vs. private
	8. Eliminated SSRowSetAdapter and SSConnectionEditor
	9. Removed Bean property code from com.nqadmin.swingSet.datasources classes
	10. Rewrote SSRowSet to extend RowSet and added default methods
	11. Rewrote SSJdbcRowSetImpl to extend JdbcRowSetImpl and implement SSRowSet
	12. Rewrote SSDBNav to add default methods from SSDBNavAdapter
	13. Deprecated SSDBNavAdapter
	14. Deprecated SSDBNavImp and replaced with SSDBNavImpl
	15. Updated @deprecated JavaDoc annotation, providing link to suggested replacement
	16. Removed SSDBCheckBox (deprecated long ago)
	17. Fixed error in SSDBNavImpl when trying to clear fields for a new record
	18. Fixes to SSDataNavigator to eliminate some redundant code and better support H2
	19. Example code cleanup and documentation.
	   
==============================================================================
SwingSet 2.2.0 - Released 2019-02-27
==============================================================================

Changes:
	1. Updated code to support Java Generics
	2. Addressed various Eclipse compilation and JavaDoc warnings
	3. Set build source & target to Java 8 (1.8)
	4. Cleaned up licensing & file header consistency using the Eclipse
	   Copyright Generator

==============================================================================
SwingSet 2.1.0 - Released 2018-09-18
==============================================================================

Changes:
	1. Fixed the issue with the last/first buttons not being enabled/disabled correctly, when navigation happens directly on the rowset.
	2. Fixed the issue of newly added item not being displayed in SSDBComboBox.
	3. Fixed NullPointerException in SSDataGrid, when users enter empty string in a date column.
	4. Fixed an exception in SSList while updating display, if the underlying rowset doesn't have any records.
	5. In SSSyncManager, updating the navigator inside combo box listener only if combo box has focus (actionPerformed is due to user action).
	6. Using locking mechanism while making any changes to the items in SSDBComboBox.
	7. Made updates/fixes to various build scripts.
	8. Updated project support email address in project docs.
	9. Added populate.sql used for sample/demo to project (formerly available only online).
	10. Corrected instructions for sample/demo.
	11. Updated GlazedLists to 1.11.0.
	12. Updated H2 used for sample/demo to 1.4.197.

==============================================================================
SwingSet 2.0.0 - Released 08-10-2012
==============================================================================

Classes or Interfaces added: 
	SSDataGridAdapter
	SSDataGridHandler
	SSList
	SSArray	
	
Changes:
	SSCheckBox:
	    1. Adding constructor that takes text for check box. Removed the preferred size setting.
	    2. Changed the way focus is transfered for ENTER key.
	
	SSComboBox:
	    1. Changed the way focus is transfered for ENTER key.
	    2. Added a function updateDisplay(boolean useTextField) which can be used in setSelectedValue.
	    3. Uses listener on rowset to update the value in the combo box. Doesn't use the listener on SSTextDocument any more. As a result removed the deprecated constructor that takes SSTextDocument.
	
	SSDataGrid:
	    1. Added SSDataGridHandler.
	    2. Disabling the use of focus traversal keys in default editor and date editor. This will allow the use of tab to move to next cell while editing.
	    3. When using a combo renderer and the underlying value is null returning index -1.
	    4. Added ability to specify if deletions should be allowed.
	    
	SSDataNavigator:
	    1. Reseting the onInsertRow flag when a new rowset is set.
	    2. Had to do conditional call to doCommitButtonClick in updatePresentRow to make sure we are on a valid row .
	    3. Added try/catch in updatePresentRow i) In updatePresentRow calling doCommitButtonClick  ii) When on insert row and allowInsertion returns false staying in insertion row rather than making a call to moveToCurrentRow().
	    4. In updatePresentRow allowUpdate was not called before calling updateRow().
	    5. Modified the code to call allowUpdate & performPostUpateOps in updatePresentRow.
	    6. While deleting the row checking for confirmDeletes variable to display the confirmation dialog.
	
	SSDBComboBox:
	    1. Added a text filter for the items in the combo box. The setFilterable() method can turn the filter on or off.
	    2. Changed the way focus is transfered for ENTER key.
	    3. In key listener ignoring function keys.
	
	SSDBNavImp:
	    1. Setting default value to null if it is an instance of SSFormattedTextField.
	

src/samples/Example5.java

/*******************************************************************************
 * Copyright (C) 2003-2019, Prasanth R. Pasala, Brian E. Pangburn, & The Pangburn Group
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. Neither the name of the copyright holder nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 * 
 * Contributors:
 *   Prasanth R. Pasala
 *   Brian E. Pangburn
 *   Diego Gil
 *   Man "Bee" Vo
 ******************************************************************************/

import java.sql.SQLException;

import javax.swing.JFrame;

import com.nqadmin.swingSet.SSDataGrid;
import com.nqadmin.swingSet.datasources.SSConnection;
import com.nqadmin.swingSet.datasources.SSJdbcRowSetImpl;

/**
 * This example demonstrates the use of an SSDataGrid to display a tabular view
 * of the suppliers & parts data.
 */
public class Example5 extends JFrame {

	/**
	 * unique serial id
	 */
	private static final long serialVersionUID = -5126011569315467420L;
	
	/**
	 * declarations
	 */
	SSConnection ssConnection = null;
	SSJdbcRowSetImpl rowset = null;
	SSDataGrid dataGrid = null;
	String url;

	/**
	 * Constructor for Example5
	 * 
	 * @param _url - path to SQL to create suppliers & parts database
	 */
	public Example5(String _url) {
		super("Example 5");
		this.url = _url;
		setSize(MainClass.childScreenWidth, MainClass.childScreenHeight);
		init();
	}

	/**
	 * Initialize the screen & datagrid
	 */
	private void init() {

		// INTERACT WITH DATABASE IN TRY/CATCH BLOCK
			try {
			// INITIALIZE DATABASE CONNECTION AND COMPONENTS
				System.out.println("url from ex 5: " + this.url);
				this.ssConnection = new SSConnection(
						"jdbc:h2:mem:suppliers_and_parts;INIT=runscript from '" + this.url + "'", "sa", "");
				this.ssConnection.setDriverName("org.h2.Driver");
				this.ssConnection.createConnection();
				
				this.rowset = new SSJdbcRowSetImpl(this.ssConnection.getConnection());
				this.rowset.setCommand("SELECT * FROM part_data ORDER BY part_name;");
			
			// SETUP THE DATA GRID - SET THE HEADER BEFORE SETTING THE ROWSET
				this.dataGrid = new SSDataGrid();
				this.dataGrid.setHeaders(new String[] { "Part ID", "Part Name", "Color Code", "Weight", "City" });
				this.dataGrid.setSSRowSet(this.rowset);
				this.dataGrid.setMessageWindow(this);
	
			// DISABLES NEW INSERTIONS TO THE DATABASE. - NOT CURRENTLY WORKING FOR H2
				this.dataGrid.setInsertion(false);
	
			// MAKE THE PART ID UNEDITABLE
				this.dataGrid.setUneditableColumns(new String[] { "part_id" });
	
			} catch (SQLException se) {
				se.printStackTrace();
			} catch (ClassNotFoundException cnfe) {
				cnfe.printStackTrace();
			}
			
		// SETUP THE CONTAINER AND ADD THE DATAGRID
			getContentPane().add(this.dataGrid.getComponent());

		// MAKE THE JFRAME VISIBLE
			setVisible(true);
	}

}

/*
 * $Log$ Revision 1.8 2012/06/07 15:54:38 beevo Modified example for
 * compatibilty with H2 database.
 *
 * Revision 1.7 2005/02/22 15:17:31 yoda2 Removed call to setPreferredSize().
 *
 * Revision 1.6 2005/02/14 18:50:25 prasanth Updated to remove calls to
 * deprecated methods.
 *
 * Revision 1.5 2005/02/04 22:40:12 yoda2 Updated Copyright info.
 *
 * Revision 1.4 2004/11/11 15:04:38 yoda2 Using TextPad, converted all tabs to
 * "soft" tabs comprised of four actual spaces.
 *
 * Revision 1.3 2004/10/25 22:01:16 yoda2 Updated JavaDoc for new datasource
 * abstraction layer in 0.9.0 release.
 *
 * Revision 1.2 2004/10/25 19:52:12 prasanth Modified to work with new SwingSet
 * (SSConnection & SSRowSet)
 *
 * Revision 1.1 2003/12/18 20:14:43 prasanth Initial commit.
 *
 */
